Firebase/Core/FIRApp.m

```diff
@@ -400,6 +401,41 @@ + (BOOL)isDefaultAppConfigured {
   return (sDefaultApp != nil);
 }
 
++ (void)registerLibrary:(nonnull NSString *)library withVersion:(nonnull NSString *)version {
+  // Create the set of characters which aren't allowed, only if this feature is used.
+  static NSCharacterSet *disallowedSet;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    NSMutableCharacterSet *allowedSet = [NSMutableCharacterSet alphanumericCharacterSet];
+    [allowedSet addCharactersInString:@"-_."];
+    disallowedSet = [allowedSet invertedSet];
+  });
+  // Make sure the library name and version strings do not contain unexpected characters, and
+  // add the name/version pair to the dictionary.
+  if ([library rangeOfCharacterFromSet:disallowedSet].location == NSNotFound &&
+      [version rangeOfCharacterFromSet:disallowedSet].location == NSNotFound) {
+    if (!sLibraryVersions) {
+      sLibraryVersions = [[NSMutableDictionary alloc] init];
+    }
+    sLibraryVersions[library] = version;
+  } else {
+    FIRLogError(kFIRLoggerCore, @"I-COR000027",
+                @"The library name (%@) or version number (%@) contain illegal characters. "
+                @"Only alphanumeric, dash, underscore and period characters are allowed.",
+                library, version);
+  }
+}
+
++ (NSString *)firebaseUserAgent {
+  NSMutableArray<NSString *> *libraries =
+      [[NSMutableArray<NSString *> alloc] initWithCapacity:sLibraryVersions.count];
+  for (NSString *libraryName in sLibraryVersions) {
+    [libraries addObject:
+        [NSString stringWithFormat:@"%@/%@", libraryName, sLibraryVersions[libraryName]]];
+  }
+  return [libraries componentsJoinedByString:@" "];
```

Comment:
Does this guarantee that the user agent would be the same every time given the same set of libraries? Can we alphabetize them or something instead?

Reply:
Changed it to sort before joining. Also updated to the test to use a full string compare rather than contains.